// @bpinternal/shared/src/schemas/webchat.ts
import * as v from 'valibot'

export const webchatConfigSchema = v.object({
  clientId: v.string(),
  configuration: v.object({
    botName: v.optional(v.string()),
    botAvatar: v.optional(v.string()),
    botDescription: v.optional(v.string()),
    phone: v.optional(v.object({ title: v.optional(v.string()), link: v.optional(v.string()) })),
    email: v.optional(v.object({ title: v.optional(v.string()), link: v.optional(v.string()) })),
    website: v.optional(v.object({ title: v.optional(v.string()), link: v.optional(v.string()) })),
    termsOfService: v.optional(v.object({ title: v.optional(v.string()), link: v.optional(v.string()) })),
    privacyPolicy: v.optional(v.object({ title: v.optional(v.string()), link: v.optional(v.string()) })),
    composerPlaceholder: v.optional(v.string()),
    color: v.optional(v.string()),
    radius: v.optional(v.number()),
    themeMode: v.optional(v.picklist(['light', 'dark'])),
    variant: v.optional(v.picklist(['solid', 'soft'])),
    fontFamily: v.fallback(v.optional(v.picklist(['rubik', 'inter', 'ibm', 'fira'])), undefined),
    showPoweredBy: v.optional(v.boolean()),
    additionalStylesheet: v.optional(v.string()),
    additionalStylesheetUrl: v.optional(v.string()),
    allowFileUpload: v.optional(v.boolean()),
  }),
})

export type WebchatConfig = v.InferOutput<typeof webchatConfigSchema>['configuration']
