// @botpress/webchat/.storybook/preview.tsx
// import '../src/index.css' // This has to be imported as early as possible
import React from 'react'
import type { Preview } from '@storybook/react'
import { StylesheetProvider } from '../src/providers'
import { Container } from '../src/components'

const radiusMap = {
  none: 0,
  sm: 0.5,
  md: 1,
  lg: 2,
  xl: 4,
}

const colorMap = {
  red: '#dc2626',
  orange: '#ea580c',
  amber: '#d97706',
  yellow: '#eab308',
  lime: '#a3e635',
  green: '#22c55e',
  emerald: '#10b981',
  teal: '#14b8a6',
  cyan: '#06b6d4',
  sky: '#0ea5e9',
  blue: '#3b82f6',
  indigo: '#6366f1',
  violet: '#8b5cf6',
  purple: '#9333ea',
  fuchsia: '#d946ef',
  pink: '#ec4899',
}

/** @type { import('@storybook/react').Preview } */
const preview: Preview = {
  initialGlobals: {
    themeMode: 'light',
    fontFamily: 'inter',
    radius: 'md',
    variant: 'solid',
    color: 'blue',
  },
  globalTypes: {
    themeMode: {
      toolbar: {
        dynamicTitle: true,
        title: 'Theme',
        icon: 'mirror',
        items: ['light', 'dark'],
      },
    },
    fontFamily: {
      toolbar: {
        title: 'Font Family',
        dynamicTitle: true,
        icon: 'edit',
        items: ['rubik', 'inter', 'ibm', 'fira'],
      },
    },
    radius: {
      toolbar: {
        title: 'Radius',
        dynamicTitle: true,
        icon: 'button',
        items: Object.keys(radiusMap),
      },
    },
    variant: {
      toolbar: {
        title: 'Variant',
        dynamicTitle: true,
        icon: 'expandalt',
        items: ['solid', 'soft'],
      },
    },
    color: {
      toolbar: {
        title: 'Color',
        dynamicTitle: true,
        icon: 'paintbrush',
        items: Object.keys(colorMap),
      },
    },
  },
  parameters: {
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
  decorators: [
    (Story, context) => {
      const { themeMode, variant, radius, fontFamily, color } = context.globals
      return (
        // TODO: These css hacks should not be required
        <Container
          className="bpReset bpContainer"
          style={{ width: '400px', border: 'none', backgroundColor: 'transparent' }}
        >
          <StylesheetProvider
            {...{ themeMode, variant, fontFamily, radius: radiusMap[radius], color: colorMap[color] }}
          />
          <Story />
        </Container>
      )
    },
  ],
}

export default preview
