// @bpinternal/shared/dist/shared2.js
import { object as e, string as t, optional as i, picklist as o, fallback as a, number as r, boolean as l } from "./shared4.js";
const c = /* @__PURE__ */ e({
  clientId: /* @__PURE__ */ t(),
  configuration: /* @__PURE__ */ e({
    botName: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    botAvatar: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    botDescription: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    phone: /* @__PURE__ */ i(/* @__PURE__ */ e({ title: /* @__PURE__ */ i(/* @__PURE__ */ t()), link: /* @__PURE__ */ i(/* @__PURE__ */ t()) })),
    email: /* @__PURE__ */ i(/* @__PURE__ */ e({ title: /* @__PURE__ */ i(/* @__PURE__ */ t()), link: /* @__PURE__ */ i(/* @__PURE__ */ t()) })),
    website: /* @__PURE__ */ i(/* @__PURE__ */ e({ title: /* @__PURE__ */ i(/* @__PURE__ */ t()), link: /* @__PURE__ */ i(/* @__PURE__ */ t()) })),
    termsOfService: /* @__PURE__ */ i(/* @__PURE__ */ e({ title: /* @__PURE__ */ i(/* @__PURE__ */ t()), link: /* @__PURE__ */ i(/* @__PURE__ */ t()) })),
    privacyPolicy: /* @__PURE__ */ i(/* @__PURE__ */ e({ title: /* @__PURE__ */ i(/* @__PURE__ */ t()), link: /* @__PURE__ */ i(/* @__PURE__ */ t()) })),
    composerPlaceholder: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    color: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    radius: /* @__PURE__ */ i(/* @__PURE__ */ r()),
    themeMode: /* @__PURE__ */ i(/* @__PURE__ */ o(["light", "dark"])),
    variant: /* @__PURE__ */ i(/* @__PURE__ */ o(["solid", "soft"])),
    fontFamily: /* @__PURE__ */ a(/* @__PURE__ */ i(/* @__PURE__ */ o(["rubik", "inter", "ibm", "fira"])), void 0),
    showPoweredBy: /* @__PURE__ */ i(/* @__PURE__ */ l()),
    additionalStylesheet: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    additionalStylesheetUrl: /* @__PURE__ */ i(/* @__PURE__ */ t()),
    allowFileUpload: /* @__PURE__ */ i(/* @__PURE__ */ l())
  })
});
export {
  c as webchatConfigSchema
};
