// @bpinternal/shared/dist/shared16.js
import { ianaRegex as E, signedOffset as F, parseFloating as l, parseMillis as W, untruncateYear as Y, parseInteger as f, isUndefined as v } from "./shared13.js";
import { monthsShort as P, weekdaysLong as Z, weekdaysShort as q } from "./shared26.js";
import y from "./shared11.js";
import z from "./shared22.js";
function x(...e) {
  const t = e.reduce((n, r) => n + r.source, "");
  return RegExp(`^${t}$`);
}
function m(...e) {
  return (t) => e.reduce(
    ([n, r, o], i) => {
      const [d, s, c] = i(t, o);
      return [{ ...n, ...d }, s || r, c];
    },
    [{}, null, 1]
  ).slice(0, 2);
}
function T(e, ...t) {
  if (e == null)
    return [null, null];
  for (const [n, r] of t) {
    const o = n.exec(e);
    if (o)
      return r(o);
  }
  return [null, null];
}
function b(...e) {
  return (t, n) => {
    const r = {};
    let o;
    for (o = 0; o < e.length; o++)
      r[e[o]] = f(t[n + o]);
    return [r, null, n + o];
  };
}
const J = /(?:(Z)|([+-]\d\d)(?::?(\d\d))?)/, G = `(?:${J.source}?(?:\\[(${E.source})\\])?)?`, h = /(\d\d)(?::?(\d\d)(?::?(\d\d)(?:[.,](\d{1,30}))?)?)?/, k = /* @__PURE__ */ RegExp(`${h.source}${G}`), A = /* @__PURE__ */ RegExp(`(?:T${k.source})?`), H = /([+-]\d{6}|\d{4})(?:-?(\d\d)(?:-?(\d\d))?)?/, L = /(\d{4})-?W(\d\d)(?:-?(\d))?/, U = /(\d{4})-?(\d{3})/, B = /* @__PURE__ */ b("weekYear", "weekNumber", "weekDay"), Q = /* @__PURE__ */ b("year", "ordinal"), j = /(\d{4})-(\d\d)-(\d\d)/, C = /* @__PURE__ */ RegExp(
  `${h.source} ?(?:${J.source}|(${E.source}))?`
), K = /* @__PURE__ */ RegExp(`(?: ${C.source})?`);
function S(e, t, n) {
  const r = e[t];
  return v(r) ? n : f(r);
}
function V(e, t) {
  return [{
    year: S(e, t),
    month: S(e, t + 1, 1),
    day: S(e, t + 2, 1)
  }, null, t + 3];
}
function g(e, t) {
  return [{
    hours: S(e, t, 0),
    minutes: S(e, t + 1, 0),
    seconds: S(e, t + 2, 0),
    milliseconds: W(e[t + 3])
  }, null, t + 4];
}
function O(e, t) {
  const n = !e[t] && !e[t + 1], r = F(e[t + 1], e[t + 2]), o = n ? null : y.instance(r);
  return [{}, o, t + 3];
}
function p(e, t) {
  const n = e[t] ? z.create(e[t]) : null;
  return [{}, n, t + 1];
}
const X = /* @__PURE__ */ RegExp(`^T?${h.source}$`), _ = /^-?P(?:(?:(-?\d{1,9}(?:\.\d{1,9})?)Y)?(?:(-?\d{1,9}(?:\.\d{1,9})?)M)?(?:(-?\d{1,9}(?:\.\d{1,9})?)W)?(?:(-?\d{1,9}(?:\.\d{1,9})?)D)?(?:T(?:(-?\d{1,9}(?:\.\d{1,9})?)H)?(?:(-?\d{1,9}(?:\.\d{1,9})?)M)?(?:(-?\d{1,20})(?:[.,](-?\d{1,9}))?S)?)?)$/;
function ee(e) {
  const [t, n, r, o, i, d, s, c, R] = e, I = t[0] === "-", M = c && c[0] === "-", u = (a, w = !1) => a !== void 0 && (w || a && I) ? -a : a;
  return [
    {
      years: u(l(n)),
      months: u(l(r)),
      weeks: u(l(o)),
      days: u(l(i)),
      hours: u(l(d)),
      minutes: u(l(s)),
      seconds: u(l(c), c === "-0"),
      milliseconds: u(W(R), M)
    }
  ];
}
const te = {
  GMT: 0,
  EDT: -4 * 60,
  EST: -5 * 60,
  CDT: -5 * 60,
  CST: -6 * 60,
  MDT: -6 * 60,
  MST: -7 * 60,
  PDT: -7 * 60,
  PST: -8 * 60
};
function D(e, t, n, r, o, i, d) {
  const s = {
    year: t.length === 2 ? Y(f(t)) : f(t),
    month: P.indexOf(n) + 1,
    day: f(r),
    hour: f(o),
    minute: f(i)
  };
  return d && (s.second = f(d)), e && (s.weekday = e.length > 3 ? Z.indexOf(e) + 1 : q.indexOf(e) + 1), s;
}
const ne = /^(?:(Mon|Tue|Wed|Thu|Fri|Sat|Sun),\s)?(\d{1,2})\s(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s(\d{2,4})\s(\d\d):(\d\d)(?::(\d\d))?\s(?:(UT|GMT|[ECMP][SD]T)|([Zz])|(?:([+-]\d\d)(\d\d)))$/;
function re(e) {
  const [
    ,
    t,
    n,
    r,
    o,
    i,
    d,
    s,
    c,
    R,
    I,
    M
  ] = e, u = D(t, o, r, n, i, d, s);
  let a;
  return c ? a = te[c] : R ? a = 0 : a = F(I, M), [u, new y(a)];
}
function oe(e) {
  return e.replace(/\([^)]*\)|[\n\t]/g, " ").replace(/(\s\s+)/g, " ").trim();
}
const se = /^(Mon|Tue|Wed|Thu|Fri|Sat|Sun), (\d\d) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) (\d{4}) (\d\d):(\d\d):(\d\d) GMT$/, de = /^(Monday|Tuesday|Wedsday|Thursday|Friday|Saturday|Sunday), (\d\d)-(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-(\d\d) (\d\d):(\d\d):(\d\d) GMT$/, ie = /^(Mon|Tue|Wed|Thu|Fri|Sat|Sun) (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) ( \d|\d\d) (\d\d):(\d\d):(\d\d) (\d{4})$/;
function $(e) {
  const [, t, n, r, o, i, d, s] = e;
  return [D(t, o, r, n, i, d, s), y.utcInstance];
}
function ce(e) {
  const [, t, n, r, o, i, d, s] = e;
  return [D(t, s, n, r, o, i, d), y.utcInstance];
}
const ue = /* @__PURE__ */ x(H, A), ae = /* @__PURE__ */ x(L, A), fe = /* @__PURE__ */ x(U, A), le = /* @__PURE__ */ x(k), N = /* @__PURE__ */ m(
  V,
  g,
  O,
  p
), Se = /* @__PURE__ */ m(
  B,
  g,
  O,
  p
), xe = /* @__PURE__ */ m(
  Q,
  g,
  O,
  p
), me = /* @__PURE__ */ m(
  g,
  O,
  p
);
function he(e) {
  return T(
    e,
    [ue, N],
    [ae, Se],
    [fe, xe],
    [le, me]
  );
}
function Ae(e) {
  return T(oe(e), [ne, re]);
}
function De(e) {
  return T(
    e,
    [se, $],
    [de, $],
    [ie, ce]
  );
}
function $e(e) {
  return T(e, [_, ee]);
}
const Te = /* @__PURE__ */ m(g);
function Ee(e) {
  return T(e, [X, Te]);
}
const ge = /* @__PURE__ */ x(j, K), Oe = /* @__PURE__ */ x(C), pe = /* @__PURE__ */ m(
  g,
  O,
  p
);
function Fe(e) {
  return T(
    e,
    [ge, N],
    [Oe, pe]
  );
}
export {
  De as parseHTTPDate,
  he as parseISODate,
  $e as parseISODuration,
  Ee as parseISOTimeOnly,
  Ae as parseRFC2822Date,
  Fe as parseSQL
};
